### ISSUE TITLE
strconv: raise error if not a number

### ISSUE THREAD

**Issue by TheBrainScrambler** [vlang/v#5421]
Hello, I've just discovered the language and have read the docs. Now I want to practice, so I'm writing a little guessing game where you enter a number between 1 and 100.
I need to convert the user's input from a string to an int. Reading the docs, I've found https://modules.vlang.io/strconv.html#atoi which does what I want. BUT ... not exactly.
If I enter some random characters in my program, `strconv.atoi` will just convert it to 0 instead of rising an error where it says that the string can't be converted.
I think this is incorrect behavior, the strconv methods should be Optionals, or at least provide methods which implements this mechanisms and are Optionals.

**TL;DR**
**Expected behavior:** If I do `strconv.atoi("blablah")` it should let us handle the error with an Optional
**Current behavior:** It just returns 0 ...

**Comment by Delta456** [comment]
What's your say on this? @penguindark

**Comment by medvednikov** [comment]
`number := os.input('enter number').int()`

**Comment by medvednikov** [comment]
But I agree, atoi() should return `?int`

**Comment by TheBrainScrambler** [comment]
"number := os.input('enter number').int()"
This also returns 0 if I enter characters instead of returning an error

**Comment by Delta456** [comment]
`str.int()` will return 0 or error if it's a not a valid number right? @medvednikov

**Comment by TheBrainScrambler** [comment]
@Delta456 No, it just returns 0 if the number is invalid, like `strconv.atoi`. Test yourself by running `number := os.input('enter number').int()` and printing `number`. Enter "Blablah" for example, it will print you 0.

**Comment by Delta456** [comment]
@TheBrainScrambler I meant if we will add an optional so what will it return.

**Comment by TheBrainScrambler** [comment]
@Delta456 I don't understand, the optional should return 0 if it isn't an error ? What ?

I also discovered by using the `cli` module that there is the same behavior: it will convert my entry to 0 if it isn't a valid number, instead of returning an error. It is probably using `strconv.atoi` or the `.int()` method

**Comment by OpenSourceAnarchist** [comment]
Yeah I'm for making pretty much all user-interaction functions return an optional type. There should be no overlap between valid returns (like 0 in this case) and invalid returns (also 0 in this case). This seems like the perfect use-case of optionals.

**Comment by penguindark** [comment]
@medvednikov @Delta456 
`atoi` in C act like: `This function returns the converted integral number as an int value. If no valid conversion could be performed, it returns zero`
In V it has the same behaviour

**Comment by OpenSourceAnarchist** [comment]
@penguindark C does not have optionals, so returning 0 is the best solution within the constraints of the language. V has a feature that would lend itself perfectly for parsing "iffy" data (unknown strings) with the certainty that whatever is returned by the function, if not an error, is precisely what the developer wants. Without optionals, there is no way to know if the user typed 0 or if they typed nonsense.

**Comment by JalonSolov** [comment]
It would perhaps be better to create a different method with a new name, instead of changing the way atoi works.

The official documentation for the parameter to atoi (in C, C++, etc.) is:

C-string **beginning with** the representation of an integral number.

The key part (that I bolded) is "beginning with".

This also means that atoi will return 5 if you pass it a string of "5ive", instead of giving an error.

So having a separate _strict_ version of atoi would be ok, but changing the way atoi works would just make porting existing code harder.

**Comment by penguindark** [comment]
I agree with JalonSolov
Add another method can be the best solution.

**Comment by TheBrainScrambler** [comment]
I don't know if the `strconv.atoi` and/or `.int()` come from C or V. Maybe just keep atoi for C compatibility by mentionning that in the docs and just modify the int method ... It would also be interesting to know what the `cli` module uses to convert strings to int

**Comment by JalonSolov** [comment]
Looks like .int()... example line:

`return flag.value.int()`

**Comment by TheBrainScrambler** [comment]
Then I think the .int() method should be modified to return an Optional and keep strconv.atoi for C compatibility.

This should also be marked as a bug. I can't properly use the CLI library because if an user enter some string on an int flag, it won't raise an error and instead return 0. But 0 is also a valid entry so I can't check if the entry is 0 ...

**Comment by medvednikov** [comment]
Yeah good point

**Comment by mcastorina** [comment]
`strconv.atoi` returns `?int` (V 0.2.2 6e589e2)

Can this issue be closed?
